pyroomacoustics/directivities.py

```diff
@@ -0,0 +1,183 @@
+
+
+class DirectivityPattern(Options):
+    FIGURE_EIGHT = "figure_eight"
+    HYPERCARDIOID = "hypercardioid"
+    CARDIOID = "cardioid"
+    SUBCARDIOID = "subcardioid"
+    OMNI = "omni"
+
+
+PATTERN_TO_CONSTANT = {
+    DirectivityPattern.FIGURE_EIGHT: 0,
+    DirectivityPattern.HYPERCARDIOID: 0.25,
+    DirectivityPattern.CARDIOID: 0.5,
+    DirectivityPattern.SUBCARDIOID: 0.75,
+    DirectivityPattern.OMNI: 1.0
+}
+
+
+class DirectionVector(object):
+    """
+    Parameters
+    ----------
+    azimuth : float
+    colatitude : float, optional
+        Default to PI / 2, only XY plane.
+    degrees : bool
+        Whether provided values are in degrees (True) or radians (False).
+    """
+    def __init__(self, azimuth, colatitude=None, degrees=True):
+        if degrees is True:
+            azimuth = np.radians(azimuth)
+            if colatitude is not None:
+                colatitude = np.radians(colatitude)
+        self._azimuth = azimuth
+        if colatitude is None:
+            colatitude = np.pi / 2
+        assert colatitude <= np.pi and colatitude >= 0
+        self._colatitude = colatitude
+
+    def get_azimuth(self, degrees=False):
+        if degrees:
+            return np.degrees(self._azimuth)
+        else:
+            return self._azimuth
+
+    def get_colatitude(self, degrees=False):
+        if degrees:
+            return np.degrees(self._colatitude)
+        else:
+            return self._colatitude
+
+
+class Directivity(object):
+    """
+
+    Parameters
+    ----------
+    orientation : DirectionVector
+        Indicates direction of the pattern.
+    pattern : str
+        One of directivities in `DirectivityPattern`.
+        TODO : support arbitrary directivities.
+
+    """
+    def __init__(self, orientation, pattern):
+        assert pattern in DirectivityPattern.values()
+        assert isinstance(orientation, DirectionVector)
+        self._orientation = orientation
+        self._pattern = pattern
+        self._p = PATTERN_TO_CONSTANT[pattern]
+
+    def get_directivity_pattern(self):
+        return self._pattern
+
+    def get_azimuth(self, degrees=True):
+        return self._orientation.get_azimuth(degrees)
+
+    def get_colatitude(self, degrees=True):
+        return self._orientation.get_colatitude(degrees)
+
+    def get_response(self, direction):
+        """
+        Get response for a single direction.
+
+        TODO : vectorize
+
+        Parameters
+        ----------
+        direction : DirectionVector
+            Direction for which to compute gain
+
+        """
+        assert isinstance(direction, DirectionVector)
+        if self._pattern == DirectivityPattern.OMNI:
+            return 1
+        else:
+            # inspiration from Habets: https://github.com/ehabets/RIR-Generator/blob/5eb70f066b74ff18c2be61c97e8e666f8492c149/rir_generator.cpp#L111
+            # we use colatitude instead of elevation
+            gain = np.sin(self._orientation.get_colatitude()) * \
```

Comment:
This may be more efficient in the vector form. Instead of using azimuth/colatitude, we store the unit vector pointing  in the direction and then we only need one inner product. This way, cos/sin only need to be called once in the constructor.

Reply:
I've switched to computing the response with Cartesian coordinates so now it is vectorized.